src/coreclr/jit/emitxarch.cpp

```diff
@@ -7055,12 +7055,170 @@ void emitter::emitIns_S(instruction ins, emitAttr attr, int varx, int offs)
     emitAdjustStackDepthPushPop(ins);
 }
 
+//----------------------------------------------------------------------------------------
+// IsRedundantMov_S_R:
+//    Check if the current `mov` instruction is redundant and can be omitted when dealing with Load/Store from stack.
+//    A `mov` is redundant in following 2 cases:
+//
+//    1. Move that is identical to last instruction emitted.
+//
+//         vmovapd  xmmword ptr [V01 rbp-20H], xmm0  # <-- last instruction
+//         vmovapd  xmmword ptr [V01 rbp-20H], xmm0  # <-- current instruction can be omitted.
+//
+//    2. Opposite Move as that of last instruction emitted.
+//
+//         vmovupd  ymmword ptr[V01 rbp-50H], ymm0  # <-- last instruction
+//         vmovupd  ymm0, ymmword ptr[V01 rbp-50H]  # <-- current instruction can be omitted.
+//
+// Arguments:
+//                 ins  - The current instruction
+//                 fmt  - The current format
+//                 size - Operand size of current instruction
+//                 ireg - The current source/destination register
```

Comment:
How does this differentiate `source` (store) vs `destination` (load)?

Reply:
> How does this differentiate `source` (store) vs `destination` (load)?

Since both register and var can be either the source or destination, we can differentiate only using instruction format - IF_SWR_RRD vs IF_RWR_SRD. If SWR - we know it's writing to 'stack' and reading from register and vice versa. The logic to handle that is from line 7181

Reply:
> Looks to be just based on `insFormat`.
> 
> Should we be asserting that the format is an expected one?

I'm skipping the optimization if format is not expected -

` if (!((emitLastIns->idInsFmt() == IF_SWR_RRD) || (emitLastIns->idInsFmt() == IF_RWR_SRD)))
    {
        return false;
    }`

Reply:
That's for the last ins format, not for the current instruction format.

I was thinking we likely should have a simple `assert((fmt == IF_SWR_RRD) || (fmt == IF_RWR_SRC))`

Reply:
Ah. I see. Will do